implementation/src/main/java/io/smallrye/mutiny/operators/multi/MultiConcatMapOp.java

```diff
@@ -50,203 +50,224 @@ public void subscribe(MultiSubscriber<? super O> subscriber) {
+        private final InnerSubscriber innerSubscriber = new InnerSubscriber();
+        private final ReentrantLock stateLock = new ReentrantLock();
+        private volatile Throwable failure;
+        private Flow.Subscription mainUpstream;
+        private volatile Flow.Subscription innerUpstream;
 
-        ConcatMapMainSubscriber(
-                MultiSubscriber<? super O> downstream,
-                Function<? super I, ? extends Publisher<? extends O>> mapper,
-                boolean delayError) {
-            this.downstream = downstream;
+        private MainSubscriber(Function<? super I, ? extends Publisher<? extends O>> mapper, boolean postponeFailurePropagation,
+                MultiSubscriber<? super O> downstream) {
             this.mapper = mapper;
-            this.delayError = delayError;
-            this.inner = new ConcatMapInner<>(this);
-        }
-
-        @Override
-        public void request(long n) {
-            if (n > 0) {
-                if (state.compareAndSet(STATE_NEW, STATE_READY)) {
-                    upstream.request(1);
-                }
-                if (deferredUpstreamRequest.compareAndSet(true, false)) {
-                    upstream.request(1);
-                }
-                inner.request(n);
-                if (inner.requested() != 0L && deferredUpstreamRequest.compareAndSet(true, false)) {
-                    upstream.request(1);
-                }
-            } else {
-                downstream.onFailure(Subscriptions.getInvalidRequestException());
-            }
-        }
-
-        @Override
-        public void cancel() {
-            while (true) {
-                int state = this.state.get();
-                if (state == STATE_CANCELLED) {
-                    return;
-                }
-                if (this.state.compareAndSet(state, STATE_CANCELLED)) {
-                    if (state == STATE_OUTER_TERMINATED) {
-                        inner.cancel();
-                    } else {
-                        inner.cancel();
-                        upstream.cancel();
-                    }
-                    return;
-                }
-            }
+            this.postponeFailurePropagation = postponeFailurePropagation;
+            this.downstream = downstream;
         }
 
         @Override
-        public void onSubscribe(Subscription subscription) {
-            if (UPSTREAM_UPDATER.compareAndSet(this, null, subscription)) {
+        public void onSubscribe(Flow.Subscription subscription) {
+            if (STATE_UPDATER.compareAndSet(this, State.INIT, State.READY)) {
+                mainUpstream = subscription;
                 downstream.onSubscribe(this);
+            } else {
+                subscription.cancel();
             }
         }
 
         @Override
         public void onItem(I item) {
-            if (!state.compareAndSet(STATE_READY, STATE_EMITTING)) {
+            if (state == State.DONE) {
```

Comment:
Note to myself: the next compare & swap covers this, we could remove this check.